BB2-612: Allow MBI/HICN Updates with FHIR_ID Matching

dtisza1 opened the pull request:
<!--
You've got a Pull Request you want to submit? Awesome!
This PR template is here to help ensure you're setup for success:
  please fill it out to help ensure that your PR is complete and ready for approval.
-->

**JIRA Ticket:**
[BB2-612](https://jira.cms.gov/browse/BB2-612)


**User Story or Bug Summary:**
<!-- Please copy-paste the brief user story or bug description that this PR is intended to address. -->
As a developer, I need the BB2 API to allow for MBI and HICN updates for beneficiaries as long as the FHIR_ID still matches so that beneficiary profiles don't encounter errors.

This updates the beneficiary matching logic to allow for changes to MBI and/or HICN values in the user-info returned from SLSx. Currently, for a beneficiary that already connected with BB2 they will receive an error, if these values have been changed/updated. From our log analysis, there are a small (but significant) number of these cases occurring.

*** MBI/HICN changes are only allowed, if the SLSX `subject/username` (primary key) and BFD `fhir_id` (patient id) remain the same! ***

Description of the types of Crosswalk changes allowed (these are logged using the following booleans for auditing purposes):

  - `mbi_updated_from_null = True` - A NULL `mbi_hash` was updated to a new `mbi_hash` value.
  - `mbi_updated = True` - The `mbi_hash` has changed and was updated to the new value from SLSx.
  - `hicn_updated` - The `hicn_hash` has changed and was updated to the new value from SLSx.

When the allowed type of changes are made to the existing `Crosswalk` record, a `CrosswalkArchive` record is created to audit the changes.

NOTE: MIGRATIONS ARE NEEDED!  This adds a new CrosswalkArchive model.


### What Does This PR Do?

<!--
Add detailed description & discussion of changes here.
The contents of this section should be used as your commit message (unless you merge the PR via a merge commit, of course).

Please follow standard Git commit message guidelines:
* First line should be a capitalized, short (50 chars or less) summary.
* The rest of the message should be in standard Markdown format, wrapped to 72 characters.
* Describe your changes in imperative mood, e.g. "make xyzzy do frotz" instead of "[This patch] makes xyzzy do frotz" or "[I] changed xyzzy to do frotz", as if you are giving orders to the codebase to change its behavior.
* List all relevant Jira issue keys, one per line at the end of the message, per: <https://confluence.atlassian.com/jirasoftwarecloud/processing-issues-with-smart-commits-788960027.html>.

Reference: <https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project>.
-->
**The following is a list of changes from the commits (including additional comments):**

* Update logging for readability
  - Change naming of `err_msg` to `mesg`, since this is not always an error type of message for `get_and_update_user()` and `create_beneficiary_record()` functions in apps/mymedicare_cb/models.py
  - Add crosswalk_updated log field for `log_get_and_update_user()` function in apps/mymedicare_cb/loggers.py.  This field has values for:  "" = No crosswalk updates, "N" = `mbi_hash` was updated from previously NULL. And to be added later:  "M" = `mbi_hash` was changed, "H" = `hicn_hash` was changed.
  - Update code to use new `crosswalk_updated` log field for `get_and_update_user()` function in apps/mymedicare_cb/models.py

  - Remove unneeded extra log entry for "UPDATE user_id_type as it has changed from the previous lookup value", since this is already reflected by the log field `hash_lookup_type` for `get_and_update_user()` function in apps/mymedicare_cb/models.py
  - Update `MYMEDICARE_CB_GET_UPDATE_BENE_LOG_SCHEMA` to require `crosswalk_updated` field for log schema in ./apps/logging/tests/audit_logger_schemas.py

* Update hicn/mbi_hash setters in Crosswalk model to allow changes
  - Update `mbi_hash.setter` and `hicn_hash.setter` to allow value changes for `Crosswalk` model class in apps/fhir/bluebutton/models.py.  Note that the `fhir_id.setter` still does not allow changes!
  - Remove `ValidationError` asserts for `hicn_hash` and `mbi_hash` changes for `test_crosswalk_setter_properties()` in apps/fhir/bluebutton/tests/test_models.py
  - Rename `test_immuatble_user_hicn_hash()` to `test_muatble_user_hicn_hash()` and remove `ValidationError` assert for `cw.user_hicn_hash` in apps/fhir/bluebutton/tests/test_models.py
  - Rename `test_immuatble_user_mbi_hash()` to `test_muatble_user_mbi_hash()` and remove `ValidationError` assert for `cw.user_hicn_hash` in apps/fhir/bluebutton/tests/test_models.py

* Update logic in `get_and_update_user()` for hicn/mbi changes
  - Move `fhir_id` can not change conditional to top as first check for existing user in crosswalk section for `get_and_update_user()` in apps/mymedicare_cb/models.py
  - Remove exception when "Found user's hicn did not match" and append an "H" to `crosswalk_updated` to provide logging info about the `hicn_hash` change to the `Crosswalk` entry for `get_and_update_user()` in apps/mymedicare_cb/models.py
  - Remove exception when "Found user's mbi did not match" and append an "M" (for mbi change) or "N" (for mbi changed from previous Null) to `crosswalk_updated` to provide logging info about the `hicn_hash` change to the `Crosswalk` entry for `get_and_update_user()` in apps/mymedicare_cb/models.py
  - Add conditional code to update the crosswalk `hash_lookup_type`, `user_mbi_hash`, `user_hicn_hash` if any of these have changed from the previously stored values for `get_and_update_user()` in apps/mymedicare_cb/models.py
  - Remove `status.HTTP_500_INTERNAL_SERVER_ERROR` assert response tests for when hicn/mbi_hash changes are made for `test_callback_exceptions()` in apps/mymedicare_cb/tests/test_callback_slsx.py.
  - Change `MyMedicareSLSxBlueButtonClientApiUserInfoTest` class to derive from `BaseApiTest` instead of `TestCase` base, and also add a `_get_log_lines_list()` method to retrive log lines for validating log entries in apps/mymedicare_cb/tests/test_callback_slsx.py.
  - Create `test_callback_allow_slsx_changes_to_hicn_and_mbi()` test method for allowed MBI and HICN updates (as long as the fhir_id still matches) in the `MyMedicareSLSxBlueButtonClientApiUserInfoTest` class in apps/mymedicare_cb/tests/test_callback_slsx.py This tests changes to HICN, MBI and both HICN & MBI are successful.

* Create and utilize CrosswalkArchive model 
  - Create ArchivedCrosswalk model in apps/fhir/bluebutton/models.py
  - Create migrations file for `ArchivedCrosswalk` model in apps/fhir/bluebutton/migrations/0007_archivedcrosswalk.py
  - Add create of a new `ArchivedCrosswalk` model instance to save an audit copy of pre-save values when there are changes to the `Crosswalk` instance for `get_and_update_user()` in apps/mymedicare_cb/models.py
  - Update tests for the `test_callback_allow_slsx_changes_to_hicn_and_mbi()` method of the `MyMedicareSLSxBlueButtonClientApiUserInfoTest` class in apps/mymedicare_cb/tests/test_callback_slsx.py to check for `ArchivedCrosswalk` entries.
  - Create admin view for `ArchivedCrosswalk` called `ArchivedCrosswalkAdmin` in apps/fhir/bluebutton/admin.py 

* Update ArchivedCrosswalk model per PR review
  - Add `username` and `date_created` to ArchivedCrosswalk in apps/fhir/bluebutton/models.py
  - Remove `crosswalk` foreign key from ArchivedCrosswalk in apps/fhir/bluebutton/models.py

* Update ArchivedCrosswalkAdmin & migrations for PR review changes
  - Create new migrations apps/fhir/bluebutton/migrations/0002_archivedcrosswalk.py
  - Add `username` to ArchivedCrosswalkAdmin in apps/fhir/bluebutton/admin.py
  - Remove `get_crosswalk_id` from ArchivedCrosswalkAdmin in apps/fhir/bluebutton/admin.py

* Create a `ArchivedCrosswalk.create()` method
  - Create a `create()` method for the `ArchivedCrosswalk` class that takes a `Crosswalk` instance as an input parameter in apps/fhir/bluebutton/models.py
  - Refactor to utilize the `ArchivedCrosswalk.create()` method for archiving the `Crosswalk` instance in the `get_and_update_user()` function in apps/mymedicare_cb/models.py
  - Add a `transaction.atomic()` context manager around `Crosswalk` archive and update code in the `get_and_update_user()` function in apps/mymedicare_cb/models.py

* Update `log_get_and_update_user()` function
  - Add `user_id` and `username` fields to the `mymedicare_cb:get_and_update_user` type log event in apps/mymedicare_cb/loggers.py
  - Update tests for new log fields added for `MyMedicareSLSxBlueButtonClientApiUserInfoTest.test_callback_allow_slsx_changes_to_hicn_and_mbi()` in apps/mymedicare_cb/tests/test_callback_slsx.py.
  - Add TODO comment related to refactoring validation logic code placement in to the `OAuth2ConfigSLSx.get_user_info()` method in apps/mymedicare_cb/views.py.

* Update loging tests and add log fields
  - Add `crosswalk_before` to log_dict when there are Crosswalk changes in the `get_and_update_user()` function in apps/mymedicare_cb/models.py
  - Move`crosswalk` log code to after Crosswalk changes in the `get_and_update_user()` function in apps/mymedicare_cb/models.py
  - Update tests to assert values for `crosswalk_before` log entries in `MyMedicareSLSxBlueButtonClientApiUserInfoTest.test_callback_allow_slsx_changes_to_hicn_and_mbi()` in apps/mymedicare_cb/tests/test_callback_slsx.py
  - Refactor logging asserts to use JSON SCHEMA for `MyMedicareSLSxBlueButtonClientApiUserInfoTest.test_callback_allow_slsx_changes_to_hicn_and_mbi()` in apps/mymedicare_cb/tests/test_callback_slsx.py
  - Update `MYMEDICARE_CB_GET_UPDATE_BENE_LOG_SCHEMA` json schema used for log validation in apps/logging/tests/audit_logger_schemas.py

* Refactor to use updated booleans and add related tests
  - Refactor `crosswalk_updated` to use `hicn_update`, `mbi_updated`, and `mbi_updated_from_null` booleans in the `get_and_update_user()` function in apps/mymedicare_cb/models.py
  - Update tests to check updated booleans for `MyMedicareSLSxBlueButtonClientApiUserInfoTest.test_callback_allow_slsx_changes_to_hicn_and_mbi()` in apps/mymedicare_cb/tests/test_callback_slsx.py
  - Update `MYMEDICARE_CB_GET_UPDATE_BENE_LOG_SCHEMA` json schema to use updated booleans in apps/logging/tests/audit_logger_schemas.py
  - Add test for `mbi_updated_from_null` case for method `MyMedicareSLSxBlueButtonClientApiUserInfoTest.test_callback_allow_slsx_changes_to_hicn_and_mbi()` in apps/mymedicare_cb/tests/test_callback_slsx.py
  - Add `MockUrlSLSxResponses.slsx_user_info_empty_hicn_mock` method for case where mbi is empty in apps/mymedicare_cb/tests/mock_url_responses_slsx.py

### What Should Reviewers Watch For?

<!--
Add some items to the following list, or remove the entire section if it doesn't apply for some reason.

Common items include:
* Is this likely to address the goals expressed in the user story?
* Are any additional documentation updates needed?
* Are there any unhandled and/or untested edge cases you can think of?
* Is user input properly sanitized & handled?
* Does this make any backwards-incompatible changes that might break end user clients?
* Can you find any bugs if you run the code locally and test it manually?
-->
If you're reviewing this PR, please check these things, in particular:

* Are there any unhandled and/or untested edge cases you can think of?
* Does this make any backwards-incompatible changes that might break end user clients?
* Can you find any bugs if you run the code locally and test it manually?
* Naming in general? Do the namings of variables, methods, classes, and other items make sense? Are they too verbose, or not verbose enough? Any redundancies? Are the comments descriptive enough?
* Are audit logging items properly handled? Is sensitive information either not included or hashed?

Example test procedure using the MSLS (mock SLSx service):

* Setup local development to use MSLS, if needed:
  - Edit docker-compose.yml file to comment the TEST SLSx ENV settings and uncomment the MSLS related ones.
  - Run:  $ docker-compose down
  - Run:  $ docker-compose up -d
  - Run:  $ docker-compose exec web docker-compose/migrate.sh
  - Launch the BB2 web service using the docker_run_web_with_keybase_secrets.sh script.

* Open the built-in test client via a browser at:  http://localhost:8000/testclient/

* Try the following tests for changes to MBI and HICN values using a new authorization flow each time. Use "testuser33" as the "SUB(username)" field for each test. Except for the fields below, the others can be left blank for these tests. Review the logging entry for the "type": "mymedicare_cb:get_and_update_user" after each test.
  - Bene's first connection to BB2 that only had a HICN value returned from SLSx:
    - HICN = "1000095313", MBI = "" (empty)
    - Expected in the logs:  "user_id_type": "H"
    - NOTE: Before MBI was added as an ID, we have older crosswalk entries where the mbi_hash = Null. This is a test of that case.
  - The bene authorizes after the MBI id was implemented.
    - HICN = "1000095313", MBI = "6S30A00AA00"
    - Expected in log:  "crosswalk_updated": "N" and "user_id_type": "M"
    - NOTE: The mbi_hash is used here to match to a fhir_id via the BFD Patient resource call.
  - The bene authorizes after their HICN id was updated in SLSx due to a PII leak of their HICN id.
    - HICN = "9000095313", MBI = "6S30A00AA00"
    - Expected in log:  "crosswalk_updated": "H" and "user_id_type": "M"
    - NOTE: The mbi_hash is used for the primary lookup here. The HICN value was updated in the crosswalk.
  - The bene authorizes after their MBI id was updated in SLSx due to a PII leak of their MBI id. We set the HICN back to the original so that a successful FHIR_ID match via the BFD is successful.
    - HICN = "1000095313", MBI = "9S30A00AA99"
    - Expected in log:  "crosswalk_updated": "HM" and "user_id_type": "H"
    - NOTE: Both the hicn_hash and mbi_hash changed in the crosswalk. The hicn_hash was used to match the fhir_id via the BFD. The primary search for mbi_hash did not match in the BFD. This case could occur where the MBI was updated in SLSx, but has not yet been updated in the BFD. 
  - The bene authorizes after their MBI id was mistakenly updated with another bene's MBI id in SLSx.
    - HICN = "1000095313", MBI = "7S30A00AA00"
    - Expected error: "Found user's fhir_id did not match" 
  - The bene authorizes after their HICN id was mistakenly updated with another bene's HICN id in SLSx. MBI is left blank so that the lookup uses hicn_hash to search via the BFD.
    - HICN = "1000082635", MBI = ""
    - Expected error: "Found user's fhir_id did not match" 

* Review the ArchivedCrosswalk entries via the Django ADMIN related to the tests. This is under the BLUEBUTTON->Archived crosswalks section. LINK: http://localhost:8000/admin/bluebutton/archivedcrosswalk/

* Using MSLS, try repeating the types of tests performed in the apps/mymedicare_cb/tests/test_callback_slsx.py `test_callback_allow_slsx_changes_to_hicn_and_mbi` unit tests.

* Can you think of more tests to try out?


### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then here's a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence: N/A.
    * Does this PR add any new software dependencies?  **No**.
    * Does this PR modify or invalidate any of our security controls? **No**.
    * Does this PR store or transmit data that was not stored or transmitted before? **No**.
* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR should not be merged unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons? **No**.

There should not be risks related to this code change.  Especially since the related code is immediately after an existing crosswalk is found for the SLSx SUB(username) and the check for the FHIR_ID not changing is immediately afterward. Therefore, any changes with the HICN or MBI values must still make a match to the same FHIR_ID via the Patient resource lookup in the BFD---before the values are updated in the crosswalk.


### What Needs to Be Merged and Deployed Before this PR?

<!--
Add some items to the following list, or remove the entire section if it doesn't apply.

Common items include:
* Database migrations (which should always be deployed by themselves, to reduce risk).
* New features in external dependencies (e.g. BFD).
-->

NOTE: MIGRATIONS ARE NEEDED!  This adds a new CrosswalkArchive model.

The following migrations should be run via the "migrate" Django manage command:

```
Migrations for 'bluebutton':
  apps/fhir/bluebutton/migrations/0002_archivedcrosswalk.py
    - Create model ArchivedCrosswalk
```


### Submitter Checklist

<!--
Helpful hint: if needed, Git allows you to edit your PR's commits and history, prior to merge.
See these resources for more information:

* <https://dev.to/maxwell_dev/the-git-rebase-introduction-i-wish-id-had>
* <https://raphaelfabeni.com/git-editing-commits-part-1/>
-->

I have gone through and verified that...:

* [ X] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [ X] I have named this PR and its branch such that they'll be automatically be linked to the (most) relevant Jira issue, per: <https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html>.
* [X ] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [X ] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [X ] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [X ] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. BFD, SLS, etc.).
* [X ] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

dtisza1 commented on apps/bb2_tools/migrations/0002_v2user.py:
```diff
@@ -0,0 +1,31 @@
+# Generated by Django 2.2.24 on 2021-07-02 17:59
```
NOTE: This migration file is related to the merged in BB2-146 migrations squash PR. So including it here.

nbragdon commented on apps/fhir/bluebutton/migrations/0002_archivedcrosswalk.py:
```diff
@@ -0,0 +1,26 @@
+                ('user_id_type', models.CharField(choices=[('H', 'HICN'), ('M', 'MBI')], default='H', max_length=1, verbose_name='Hash ID type last used for FHIR_ID lookup')),
+                ('_user_id_hash', models.CharField(db_column='user_id_hash', db_index=True, default=None, max_length=64, verbose_name='HASH of User HICN ID')),
+                ('_user_mbi_hash', models.CharField(db_column='user_mbi_hash', db_index=True, default=None, max_length=64, null=True, verbose_name='HASH of User MBI ID')),
+                ('archived_at', models.DateTimeField(auto_now_add=True)),
```
Do you think we should go with the more generic database column names of: `updatedAt`, `createdAt`. Those can basically apply to any table and will give the same thing we are going for here

dtisza1 replied:
> Do you think we should go with the more generic database column names of: `updatedAt`, `createdAt`. Those can basically apply to any table and will give the same thing we are going for here

Was keeping the naming similar to the other Archived* type models.

For example: 

https://github.com/CMSgov/bluebutton-web-server/blob/fe3ac84a83b37528d75c37674eddd6e389b85712/apps/dot_ext/models.py#L223
https://github.com/CMSgov/bluebutton-web-server/blob/7e33b3e557cf0159740515cdd9405b359fe92151/apps/authorization/models.py#L46

However, I am missing a copy of the `date_created` field from the Crosswalk instance.

I like the idea of us standardizing on `created_at` and `updated_at` for our models. There looks to be a mix of naming for those across our models.

I'll update the model shortly.

nbragdon replied:
Also don't think you need it but this link is super small and just shows the ez commands for those 2 fields:
https://timonweb.com/django/automatic-created-and-updated-datetime-fields-for-django-models/

nbragdon commented on apps/fhir/bluebutton/migrations/0002_archivedcrosswalk.py:
```diff
@@ -0,0 +1,26 @@
+                ('_user_id_hash', models.CharField(db_column='user_id_hash', db_index=True, default=None, max_length=64, verbose_name='HASH of User HICN ID')),
+                ('_user_mbi_hash', models.CharField(db_column='user_mbi_hash', db_index=True, default=None, max_length=64, null=True, verbose_name='HASH of User MBI ID')),
+                ('archived_at', models.DateTimeField(auto_now_add=True)),
+                ('crosswalk', models.ForeignKey(blank=True, db_constraint=False, null=True, on_delete=django.db.models.deletion.CASCADE, to='bluebutton.Crosswalk')),
```
Is the foreign key necessary you think?

dtisza1 replied:
> Is the foreign key necessary you think?

You're right. This is not needed. Will update the code shortly.

nbragdon replied:
Thinking on this more - it's basically if we want the archive to outlast the crosswalk record. Like if the crosswalk record is deleted, do we want the archived/audit trail to still exist for it? I'm thinking yes? Maybe we'll create some nightly job to eventually clean up the archived table but for now I think keeping the data around is good.

dtisza1 replied:
> Thinking on this more - it's basically if we want the archive to outlast the crosswalk record. Like if the crosswalk record is deleted, do we want the archived/audit trail to still exist for it? I'm thinking yes? Maybe we'll create some nightly job to eventually clean up the archived table but for now I think keeping the data around is good.

Yes to this :+1:

nbragdon commented on apps/mymedicare_cb/models.py:
```diff
@@ -57,43 +57,52 @@ def get_and_update_user(subject, mbi_hash, hicn_hash, first_name, last_name, ema
 
+        # Did the hicn change?
+        if user.crosswalk.user_hicn_hash != hicn_hash:
+            crosswalk_updated = crosswalk_updated + "H"
```
Couple of notes here:
Do you think we could just make it an array of full name values? Just makes to easier to read. ex: `["HICN", "MBI"]`
Also for all the values, can we just go ahead and create constants? ie: 
```
HICN_UPDATED_TEXT = "HICN"

crosswalk_updated = crosswalk_updated + [HICN_UPDATED_TEXT]
```

nbragdon replied:
Another option could be creating boolean values for each:
```
hicn_updated: true
mbi_updated: false
```

(Then you wont need the 3rd option because you can infer from those two)

nbragdon commented on apps/mymedicare_cb/models.py:
```diff
@@ -57,43 +57,52 @@ def get_and_update_user(subject, mbi_hash, hicn_hash, first_name, last_name, ema
+        # Update Crosswalk if there are any allowed changes or hash_type used for lookup changed.
+        if user.crosswalk.user_id_type != hash_lookup_type or crosswalk_updated != "":
+            # Copy pre-save values to ArchivedCrosswalk instance
+            acw = ArchivedCrosswalk.objects.create(crosswalk=user.crosswalk,
```
Think it'd be worth having a convenience function for this?
Something like:
`acw = ArchivedCrosswalk.objects.create(crosswalk)`
Where you just pass another crosswalk object into the function and it creates the archived crosswalk row from it?

JFU-GIT replied:
and put both updates in one TX demarcation:
 with transaction.atomic():
  .............................. update crosswalk and insert archived crosswalk.

dtisza1 replied:
> Think it'd be worth having a convenience function for this?
> Something like:
> `acw = ArchivedCrosswalk.objects.create(crosswalk)`
> Where you just pass another crosswalk object into the function and it creates the archived crosswalk row from it?

@nbragdon I like this!  Just updating the code now.  Creating a `ArchivedCrosswalk.create()` method.

dtisza1 replied:
> and put both updates in one TX demarcation:
> with transaction.atomic():
> .............................. update crosswalk and insert archived crosswalk.

@JFU-GIT Good catch and idea here!

JFU-GIT reviewed (commented):
after allowing mbi / hicn updates, the vulnerability level increased from previous implementation where mbi/hicn update not allowed and when that occur, error raised and auth aborted.

but valid use cases requires mbi/hicn updates, glad to see there are measures in place to track update activities (archived crosswalk), it would be more helpful if some active monitoring and alert enforced, e.g. if such updates occurring at an unusual rate and/or extensiveness, say, received updates on 100+ benes in 1 day..., this way, possible leakage would be going unnoticed for extensive period of time...

oragame commented on apps/fhir/bluebutton/tests/test_models.py:
```diff
@@ -78,27 +70,27 @@ def test_immutable_fhir_id(self):
             cw.fhir_id = "-20000000000002"
 
-    def test_immuatble_user_hicn_hash(self):
+    def test_muatble_user_hicn_hash(self):
```
Minor detail, but is this mutable? thinking this is just a typo. Same below on line 84.

dtisza1 replied:
> Minor detail, but is this mutable? thinking this is just a typo. Same below on line 84.

@oragame I renamed the test methods, since hicn/mbi hash values can now be changed. Before they could not.

njdister replied:
Makes sense but I think there is a typo here.

dtisza1 replied:
> Makes sense but I think there is a typo here.

@oragame Very sorry :man_facepalming: .  I read your comment too quickly and didn't see the actual typo before responding :smile:   Also didn't notice it in the original code.   Will fix shortly.   @njdister

oragame reviewed (commented):
This is looking really good imo. I did like Nick's comment as far as using `updatedAt` and `createdAt` column names to be more generic as well as allowing a reference to the date a record is created. I will let the more experienced team members determine readiness, but great work overall!

JFU-GIT reviewed (commented):
some slsx user info validations are done in apps/mymedicare_db/views.py line 61-99, would it be better encapsulation if move validations into class OAuth2ConfigSLSx(object)->get_user_info(self, request)?
no big deal and not in the changes of this PR.

JFU-GIT reviewed (commented):
current crosswalk implementation assumes communication with slsx is never compromised, to make MITM more difficult, some protocol level extension similar to OAUTH2 token introspect extension seems popular best practice, for OAUTH2 access token introspect extension, resource server introspect auth server about the access token, if it's active.

in SLSX service, a mbi check call to SLSX upon mbi update could help greatly reduce MITM risk, the request could just send the incoming mbi and the response could be the previous mbi to be replaced (bb2 check it with the incombent mbi) hence validate the authenticity of the incoming mbi OR rejects it.

dtisza1 commented:
> some slsx user info validations are done in apps/mymedicare_db/views.py line 61-99, would it be better encapsulation if move validations into class OAuth2ConfigSLSx(object)->get_user_info(self, request)?
> no big deal and not in the changes of this PR.

@JFU-GIT I agree that would be nicer/cleaner code! Maybe can include changing it as part of a future code review and refactoring story.  I will add a TODO in the comments for it now.

dtisza1 commented:
Update:
I just pushed up some changed related to the all of the good review comments.  TY everyone for those!!!

I'm working on this next before requesting re-reviews:

* Refactor to use booleans for crosswalk_updated per Nick's review comments.
* Merge w/ master (includes refactor to use new logging class)
* Refactor logging tests to utilize our newer JSONSCHEMA asserts standard.
* Push up all changes and re-request review

njdister commented:
@dtisza1 It looks like a lot of escape/regex characters leaked in as I'm seeing `^` and `$` everywhere in the tests now.

dtisza1 commented:
> @dtisza1 It looks like a lot of escape/regex characters leaked in as I'm seeing `^` and `$` everywhere in the tests now. Or is that expected?

@njdister Those are intended.  Was just cleaning up some of the json schemas used for the log testing. Hoping this doesn't make things too noisy for review.   I did this only in the areas I was working with.  So I added the regex's for exactly matching patterns.

I also added a `"type"` to each field. I found some issues with the pattern matching, if the type isn't included. For example, if a log field is a NULL rather than a string, the pattern match is successful---even though the pattern didn't match.

These type of changes should be made with the other schemas and unit tests too. Will add some notes to the related Jira story we have in the backlog.

I'm just finishing up and will add some self-review comments, like for the type of changes you noticed. Then will request re-reviews soon.

TY!

dtisza1 commented on apps/mymedicare_cb/models.py:
```diff
@@ -61,71 +61,92 @@ def get_and_update_user(
 
     log_dict = {
         "type": "mymedicare_cb:get_and_update_user",
+        "subject": subject,
```
Added this to the logging. It is the SLSx subject (unique username/primary id key) for the beneficiary.

dtisza1 commented on apps/mymedicare_cb/models.py:
```diff
@@ -61,71 +61,92 @@ def get_and_update_user(
         "hicn_hash": hicn_hash,
         "hash_lookup_type": hash_lookup_type,
         "crosswalk": {},
+        "crosswalk_before": {},
```
Added this to the logging to show what the `Crosswalk` record contained before an update.

dtisza1 commented on apps/mymedicare_cb/models.py:
```diff
@@ -61,71 +61,92 @@ def get_and_update_user(
     }
 
+    # Init for types of crosswalk updates.
+    hicn_updated = False
+    mbi_updated = False
+    mbi_updated_from_null = False
```
Refactored to use these flags to audit the type of Crosswalk changes in the logs. This is from Nick's review suggestion.

dtisza1 commented on apps/mymedicare_cb/models.py:
```diff
@@ -61,71 +61,92 @@ def get_and_update_user(
+            })
             logger.info(log_dict)
-            raise BBMyMedicareCallbackCrosswalkUpdateException(err_msg)
+            raise BBMyMedicareCallbackCrosswalkUpdateException(mesg)
```
Changed `err_msg` to `mesg` for consistency/readabiliy.

dtisza1 commented on apps/mymedicare_cb/models.py:
```diff
@@ -61,71 +61,92 @@ def get_and_update_user(
+            "mbi_updated": mbi_updated,
+            "mbi_updated_from_null": mbi_updated_from_null,
+            "mesg": "RETURN existing beneficiary record",
+            "crosswalk": {
+                "id": user.crosswalk.id,
+                "user_hicn_hash": user.crosswalk.user_hicn_hash,
+                "user_mbi_hash": user.crosswalk.user_mbi_hash,
+                "fhir_id": user.crosswalk.fhir_id,
+                "user_id_type": user.crosswalk.user_id_type,
+            },
```
Moved the location of this logging code to show the final state of the `Crosswalk`. `crosswalk_before` was added to show the state before updates are made.

dtisza1 commented on apps/mymedicare_cb/models.py:
```diff
@@ -185,85 +212,107 @@ def create_beneficiary_record(
-        err_msg = "username can not be None"
-        log_dict["status"] = "FAIL"
-        log_dict["mesg"] = err_msg
+        mesg = "username can not be None"
+        log_dict.update({
+            "status": "FAIL",
+            "mesg": mesg,
+        })
```
Just refactored to utilize `.update()`.  Probably not better for this case, but I think it's nicer when there are more dict keys involved. There are similar changes in this file.

dtisza1 commented on apps/mymedicare_cb/tests/test_callback_slsx.py:
```diff
@@ -434,3 +429,467 @@ def catchall(url, request):
+
+        # 1. First successful matching for beneficiary having valid only hicn and EMPTY mbi.
+        with HTTMock(MockUrlSLSxResponses.slsx_token_mock,
+                     MockUrlSLSxResponses.slsx_user_info_empty_mbi_mock,
```
Note `MockUrlSLSxResponses.slsx_user_info_empty_mbi_mock` difference. This will change in later tests as well, depending on what type of mock response is needed.

dtisza1 commented on apps/mymedicare_cb/tests/test_callback_slsx.py:
```diff
@@ -434,3 +429,467 @@ def catchall(url, request):
+        #   Get last log line
+        log_dict = json.loads(log_list[len(log_list) - 1])
+
+        #   Set working copy of schema
+        log_schema = copy.deepcopy(MYMEDICARE_CB_GET_UPDATE_BENE_LOG_SCHEMA)
```
Making use of a DEEP COPY of the schema so that the original schema can be used/reset as a basis for each of the tests. Changes to the `log_schema` dict do not affect the original `MYMEDICARE_CB_GET_UPDATE_BENE_LOG_SCHEMA` schema dict.

dtisza1 reviewed (commented):
These are some self review comments that I hope are helpful.

I'll re-request reviews shortly. Just pushing up some minor changes for stuff I noticed.

I left out a few unit tests related to when the hashing SALT/PEPPER is changed for the back end BFD matching.  Will test this out manually using MSLS and the number 9 test is also good for this type of change . Would like to add unit tests for this to be complete, but was running in to an issue when changing `settings.USER_ID_SALT` and `settings.USER_ID_ITERATIONS` using overrides. Found that this partially works in debugging, but the override is not working related to an imported function `get_user_id_salt` in  apps/fhir/bluebutton/models.py. Still trying to figure out :-) 

Thank you for the great PR review comments and suggestions so far!

njdister commented on apps/fhir/bluebutton/models.py:
```diff
@@ -134,22 +134,60 @@ def user_mbi_hash(self):
-            raise ValidationError("this value cannot be modified.")
-        if self._user_id_hash:
-            raise ValidationError("this value cannot be modified.")
         self._user_id_hash = value
```
During security review of the auth flow dashboard, there is still a very small percentage of beneficiaries who only have HICNs, not MBIs.  So HICN matching will still be needed for those edge cases.

njdister commented on apps/fhir/bluebutton/tests/test_models.py:
```diff
@@ -78,27 +70,27 @@ def test_immutable_fhir_id(self):
+        cw.save()
 
-    def test_immuatble_user_mbi_hash(self):
+    def test_muatble_user_mbi_hash(self):
```
Typo here as well.

JFU-GIT reviewed (commented):
noticed 2 scenarios:

1. with a good hicn provided, basically any mbi can be provided and the crosswalk record got updated with the mbi hash, e.g. changed mbi everytime from MSLSX login form, result in archived crosswalk like below:

![image](https://user-images.githubusercontent.com/25412191/126820301-34a65feb-999a-4b8a-9655-167ec86a13e9.png)
 
2. with a good matching mbi and a changed hicn (bad hicn), the algorithm just ignored the changed hicn, the current hicn (hash) remain the same in crosswalk, and the fact that a changed hicn received from slsx userinfo is just ignored silently.

should a warning or alert be logged here?

nbragdon reviewed (approved):
This is looking good to me, thanks for your good work on this @dtisza1 👍

nbragdon commented:
@JFU-GIT I just tried:
Good matching MBI
Bad non-matching HICN
The HICN updated correctly.
You want to try it again? Maybe you just didn't update your DB before checking?

JFU-GIT reviewed (approved):
this looks good except some super paranoid scenarios

njdister reviewed (approved):
Great work on this!  I did extensive testing of this on Thursday, and my only concerns were around hypothetical scenarios that aren't directly related to the scope of this PR (new benes coming in with null HICNs, etc).  No reason to hold this up for hypotheticals as this fixes a real issue we have with returning benes presently.

dtisza1 commented:
@njdister @JFU-GIT @nbragdon I just performed some local testing for backward compatible migrations---if needing to roll back to r73.  This worked!!!

Here are the steps I used:

1. Start clean with r73
```sh
$ docker-compose down
$ docker images
$ docker rmi bluebutton-web-server_unittests bluebutton-web-server_web bluebutton-web-server_msls postgres
$ git checkout r73
$ docker-compose up -d
$ docker-compose exec web docker-compose/migrate.sh
$ dev-local/docker_run_web_with_keybase_secrets.sh
```
2. Perform tests using the built-in test client using the following synth benes. This first test uses the slsx and medicare.gov login/pass config locally.
```text
BBUser00000
"patient": "-19990000000001"

BBUser00001
'patient': '-19990000000002'
```
3. Switch to testing with PR branch:
```sh
# ctrl-c to stop web service!
git checkout dtisza1/BB2-612-beneficiary-matching-logic-updates
$ docker-compose start web
$ docker-compose exec web python manage.py showmigrations 
$ docker-compose exec web python manage.py migrate # Migrations should already have been run, but do this to verify.
$ vi docker-compose.yml  # Edit for mslsx mode testing
$ cp docker-compose.yml docker-compose.yml_mslsx_version
$ dev-local/docker_run_web_with_keybase_secrets.sh
```
4. Perform tests using the built-in test client using the following synth benes. This second testing uses the mslsx config locally so that MBI values can be changed. You should see a change to using HICN for the secondary type FHIR_ID matching.  Afterward, check that archivedCrosswalk records are created in the DB/Django Admin.
```text
BBUser00000
"patient": "-19990000000001"
SLSX_SUBJECT:  0854b44c-ddf7-4f49-818e-1cf22d028d67
HICN: 1000079035
MBI: 1S00A00AA00  --> change to in test: 9S00A00AA99

BBUser00001
'patient': '-19990000000002'
SLSX_SUBJECT:  0854b44c-b078-4e9e-a01c-ab4ed6baf230
HICN: 1000003701
MBI: 2S00A00AA00  --> change to in test: 9S00A00AA98
```
5. Roll back to r73 and test that it is backward compatible.
```sh
# ctrl-c to stop web service!
$ git checkout -- docker-compose.yml
$ git checkout r73
$ cp docker-compose.yml_mslsx_version docker-compose.yml
$ clear; docker-compose exec web python runtests.py
$ dev-local/docker_run_web_with_keybase_secrets.sh
```
6. Perform tests using the built-in test client using the following synth benes. This third testing also uses the mslsx config locally. You should get errors when trying to change either HICN/MBI values from those previously used. Using the previously used values are successful here!!!
```text
BBUser00000
"patient": "-19990000000001"
SLSX_SUBJECT:  0854b44c-ddf7-4f49-818e-1cf22d028d67
HICN: 1000079035
MBI: 9S00A00AA99

BBUser00001
'patient': '-19990000000002'
SLSX_SUBJECT:  0854b44c-b078-4e9e-a01c-ab4ed6baf230
HICN: 1000003701
MBI: 9S00A00AA98
```

dtisza1 commented:
> current crosswalk implementation assumes communication with slsx is never compromised, to make MITM more difficult, some protocol level extension similar to OAUTH2 token introspect extension seems popular best practice, for OAUTH2 access token introspect extension, resource server introspect auth server about the access token, if it's active.
> 
> in SLSX service, a mbi check call to SLSX upon mbi update could help greatly reduce MITM risk, the request could just send the incoming mbi and the response could be the previous mbi to be replaced (bb2 check it with the incombent mbi) hence validate the authenticity of the incoming mbi OR rejects it.

@JFU-GIT These are some good improvement ideas. I'm capturing this for a backlog Jira story. TY!

JFU-GIT commented:
> @njdister @JFU-GIT @nbragdon I just performed some local testing for backward compatible migrations---if needing to roll back to r73. This worked!!!
> 
> Here are the steps I used:
> 
> 1. Start clean with r73
> 
> ```shell
> $ docker-compose down
> $ docker images
> $ docker rmi bluebutton-web-server_unittests bluebutton-web-server_web bluebutton-web-server_msls postgres
> $ git checkout r73
> $ docker-compose up -d
> $ docker-compose exec web docker-compose/migrate.sh
> $ dev-local/docker_run_web_with_keybase_secrets.sh
> ```
> 
> 1. Perform tests using the built-in test client using the following synth benes. This first test uses the slsx and medicare.gov login/pass config locally.
> 
> ```
> BBUser00000
> "patient": "-19990000000001"
> 
> BBUser00001
> 'patient': '-19990000000002'
> ```
> 
> 1. Switch to testing with PR branch:
> 
> ```shell
> # ctrl-c to stop web service!
> git checkout dtisza1/BB2-612-beneficiary-matching-logic-updates
> $ docker-compose start web
> $ docker-compose exec web python manage.py showmigrations 
> $ docker-compose exec web python manage.py migrate # Migrations should already have been run, but do this to verify.
> $ vi docker-compose.yml  # Edit for mslsx mode testing
> $ cp docker-compose.yml docker-compose.yml_mslsx_version
> $ dev-local/docker_run_web_with_keybase_secrets.sh
> ```
> 
> 1. Perform tests using the built-in test client using the following synth benes. This second testing uses the mslsx config locally so that MBI values can be changed. You should see a change to using HICN for the secondary type FHIR_ID matching.  Afterward, check that archivedCrosswalk records are created in the DB/Django Admin.
> 
> ```
> BBUser00000
> "patient": "-19990000000001"
> SLSX_SUBJECT:  0854b44c-ddf7-4f49-818e-1cf22d028d67
> HICN: 1000079035
> MBI: 1S00A00AA00  --> change to in test: 9S00A00AA99
> 
> BBUser00001
> 'patient': '-19990000000002'
> SLSX_SUBJECT:  0854b44c-b078-4e9e-a01c-ab4ed6baf230
> HICN: 1000003701
> MBI: 2S00A00AA00  --> change to in test: 9S00A00AA98
> ```
> 
> 1. Roll back to r73 and test that it is backward compatible.
> 
> ```shell
> # ctrl-c to stop web service!
> $ git checkout -- docker-compose.yml
> $ git checkout r73
> $ cp docker-compose.yml_mslsx_version docker-compose.yml
> $ clear; docker-compose exec web python runtests.py
> $ dev-local/docker_run_web_with_keybase_secrets.sh
> ```
> 
> 1. Perform tests using the built-in test client using the following synth benes. This third testing also uses the mslsx config locally. You should get errors when trying to change either HICN/MBI values from those previously used. Using the previously used values are successful here!!!
> 
> ```
> BBUser00000
> "patient": "-19990000000001"
> SLSX_SUBJECT:  0854b44c-ddf7-4f49-818e-1cf22d028d67
> HICN: 1000079035
> MBI: 9S00A00AA99
> 
> BBUser00001
> 'patient': '-19990000000002'
> SLSX_SUBJECT:  0854b44c-b078-4e9e-a01c-ab4ed6baf230
> HICN: 1000003701
> MBI: 9S00A00AA98
> ```

@dtisza1 thanks for the thorough work, thanks!

dtisza1 commented:
> noticed 2 scenarios:
> 
> 1. with a good hicn provided, basically any mbi can be provided and the crosswalk record got updated with the mbi hash, e.g. changed mbi everytime from MSLSX login form, result in archived crosswalk like below:
> 
> 
> 1. with a good matching mbi and a changed hicn (bad hicn), the algorithm just ignored the changed hicn, the current hicn (hash) remain the same in crosswalk, and the fact that a changed hicn received from slsx userinfo is just ignored silently.
> 
> should a warning or alert be logged here?

@JFU-GIT Was just checking out this scenario. Will capture this in a Jira backlog story to review more later.

This looks to be working as expected with updating the crosswalk with updated hicn/mbi hash values and an archive record with before values.  A good log event type showing this is `type="mymedicare_cb:get_and_update_user"` where it shows the `before_crosswalk` and `crosswalk` (after) values. 

You might have a good idea with adding some kind of warning logging for this later. We can utilize the current logging to report on cases of this happening.

dtisza1 commented:
> @JFU-GIT I just tried:
> Good matching MBI
> Bad non-matching HICN
> The HICN updated correctly.
> You want to try it again? Maybe you just didn't update your DB before checking?

@JFU-GIT @nbragdon I was seeing this working OK when testing just now related to updating the crosswalk and creating an achived crosswalk record too.

dtisza1 commented:
> **There is a scenario (super paranoid): MITM attach replace slsx userinfo response mbi with a good beneficiary's mbi (but not belong to the bene of the crosswalk), then mbi will look up the patient but fhir id does not match, exception raised, if MITM could inject that mbi for each userinfo response, then all auth will fail...**

For an existing bene in the crosswalk, any MITM changes to subject, MBI and/or HICN would need the matching fhir_id via the BFD back end to match the crosswalk.fhir_id. This throws an exception if not matching and is the first check for a returning bene (w/ existing crosswalk entry).

Another fringe case would be if the HICN was for a different user, but the MBI was good. The fhir_id would match using the MBI, but the hicn_hash would be stored in the crosswalk. This would produce an error for the other bene logging in later.  However, the hicn would never be successfully used for matching via the back end---but just recorded in the crosswalk.
